### DISCUSSION TITLE
More complete FIOS Remote and a few other mods

### DISCUSSION THREAD

**Discussion by lepmail** [GerryDazoo/Slinger#224] [General]
First of all, a BIG THANK YOU to Gerry for all his efforts!!!!  I have been using SlingBoxes for many, many years now and hope to continue to do so, thanks to Gerry and other contributors. Current setup: Solo, M2 and SB500, Verizon Fios and VLC on Windows 8.1/10 and Fire Stick.

Since Verizon pushed put a new update to the CHS435 DVRs, the boxes have been turned themselves off after 4 hours.  Therefore, I needed a power button on the Fios remote.  Using the remote file that was uploaded not long ago, I created the following:
![image](https://user-images.githubusercontent.com/40434981/199676907-9fe07104-4ec7-433c-860f-4529b0c2fb0b.png)

I did run into an issue with the special arrow characters not displaying properly in my environment, so I modified the code to accept another value to be used for "display" after the style on the button tag, like this:  **'UP' : 38 : arrow-button : &amp;#9650;**

I also added this code at in BuildPage (at line 1191) to prevent the remote from scrolling back up after you click on a button.

```
<script>
document.addEventListener('DOMContentLoaded', function() {
    var sep = '\uE000'; // an unusual char: unicode 'Private Use, First'
    window.addEventListener('pagehide', function(e) {
        window.name += sep + window.pageXOffset + sep + window.pageYOffset;
    });
    if  (window.name && window.name.indexOf(sep) > -1) {
        var parts = window.name.split(sep);
        if  (parts.length >= 3) {
            window.name = parts[0];
            window.scrollTo(parseFloat(parts[parts.length - 2]), parseFloat(parts[parts.length - 1]));
        }
    }
});
</script>
```
Lastly, I made a few other changes to the latest version (3.08e??) to:
* automatically update the action attribute in remote_xxxx.html files (<form method="post" action="/Remote">) to the corresponding Slingbox name
* allow the Digits field to accept multiple test codes after the "?" separated by commas
* allow "#"+channel number to directly tune to a channel, like **'MSNBC' : &#35; 103** , instead of 'MSNBC': 9,18,11, which might be more convenient to code

The updated slingbox_server".txt" is attached -- this editor would not allow me to upload the ".py" version.  

Although the revised remote file is attached, it will not fully render unless the "display" change is made to the code.

Gerry -- I'd appreciate your feedback/thoughts. Thanks again!

[FIOS_remote_fancy.txt](https://github.com/GerryDazoo/Slinger/files/9927136/FIOS_remote_fancy.txt)
[slingbox_server_modified.txt](https://github.com/GerryDazoo/Slinger/files/9927250/slingbox_server_modified.txt)

**Comment by GerryDazoo** [comment]
I'm losing my mind. I'd swear I responded yesterday.
The code to update the action field automatically is a good idea. Did you test it with an old config.ini and a integrated_config.ini?
What use case did you have for allowing multiple test keycodes?
Page scrolling. Hmm, Not sure I want to make that the default. We can always make a "non-scrolling" custom remote file for people to use as a starting point.
It definitely looks like you know your HTML and Java Script. Have you considered using the "raw html" feature in the latest release to make your page? There would be no code changes required.

**Reply by lepmail** [reply]
Thanks for your comments.
Yes, the action field code works with either version of the ini.
I wanted to test stacking up commands, other than channel numbers, as it takes a bit of navigation on the FIOS remote to replay recorded shows.  Thus, the need for multiple test keycodes.
I had made these customizations two weeks ago before your change to allow for an external html file. Will try that next. Have you considered pulling the default remote "html" out of the source altogether? That way you could distribute different default versions that users could start from.  Just a thought.

**Reply by GerryDazoo** [reply]
One more question. Why didn't you put the java script code into the remote.txt file instead of hardcoding it in the software?
That should work, if it doesn't I need to find out why.
There's many starting places for remotes in the CustomRemotes folder in the repository.

**Reply by lepmail** [reply]
The way you currently have BuildPage coded, only the "style" and "form" sections get substituted by what's in remote.txt.  That's why I added the java script to the server code. Amore elegant approach might be to perhaps allow for a "script=" section in remote.txt, but that could also create other issues/vulnerabilities down the line.

**Reply by GerryDazoo** [reply]
good point. I'll consider adding it in.

**Comment by theweatherguy** [comment]
Wow, that looks great.  Impressive!  Thank you.   I'm looking forward to using it.  But  I'm going wait for Gerry and you to review his questions before trying to implement it.

I got it working.  I see that all I had to do was to reference the FIOS_fancy.txt in the config.ini file and place the text file in the same directory.

**Reply by lepmail** [reply]
Thank you!  It took some time to get it to look like this -- and I am not done yet! ;)
In answer to your questions, just specify **remote=FIOS_remote_fancy.txt** in the ini.  The remote will work using Gerry's server code, but not render exactly as shown above with the special characters for the arrows, etc.
The code changes I made were suggest enhancements and not vital to the operation of this remote.
I've started the conversion to a full HTML file, but ran into a couple of formatting issues, which I hope to clean up this weekend.  Once I have that ready I will upload it again and it should render 100% with the special characters.
Off to work now, more later.

**Reply by pfuller88** [reply]
Great work.  I copied some of your button codes for the TIVO html remote i’m working on.

**Reply by lepmail** [reply]
Thanks and good luck with your TIVO remote. I have 2 old TIVO boxes, but have never hooked them up to my sling boxes.

**Reply by mdehrlich** [reply]
Did you update your remote to work with the new "e" release?

**Reply by mdehrlich** [reply]
Any progress on html version?

**Reply by lepmail** [reply]
@mdehrlich here is the html version, which needs to be renamed before use.
[FIOS_remote_fancy_html.txt](https://github.com/GerryDazoo/Slinger/files/10051075/FIOS_remote_fancy_html.txt)

**Comment by mdehrlich** [comment]
Thank you to lep for the great changes.  I'll be jumping on this for my FIOS boxes!
Made my mods to change preset channels, and change to allow 0 resolution.  It all works great!

**Reply by lepmail** [reply]
Thanks! I'm curious -- what's your change to allow 0 resolution??

**Reply by mdehrlich** [reply]
The change to allow 0 resolution (audio only) is made in the python code.  Sorry if you got the idea that it was in the remote.

**Reply by mdehrlich** [reply]
You don't need a power button.  Any (dvr) button will turn on power.  I also eliminated the volume, mute and the bottom row, because they don't have anything to do with the fios box.  Here is my new version.  With the channel input field, I don't need numbers.  So everything I need fits nicely in the upper sections.
![image](https://user-images.githubusercontent.com/8422167/200449788-03558b4c-9f11-49c8-a0fb-241bdd2590c1.png)

**Reply by RandomMan719** [reply]
> The change to allow 0 resolution (audio only) is made in the python code.  Sorry if you got the idea that it was in the remote.

Will resolution=0 be included in the next release of slingbox_server.exe

**Reply by GerryDazoo** [reply]
I was unaware that resolution=0 was a valid value. I'll give it test. Many people have been asking about audio only mode. It's a two character change to the code.
It will be in the next release.

**Reply by mdehrlich** [reply]
I made the mod a to test it.  That's how I know it works for my boxes.  Thanks for adding it to the release candidate.

**Reply by RandomMan719** [reply]
Yes, thank you Gerry.

**Comment by as400jockey** [comment]
FYI - if you have Verizon Fios, you can use the Fios TV app on your phone as a remote.  I haven't tried it from outside the house, yet, but it works inside.  

Not pretty, but it's better than the remote functionality on the Slingplayer for Android...

GERRY - this thing is AWESOME.  Got it working on my internal network with 1 of my two M1s.  Going to enjoy it a while, then add my other M1... then my Solo.  Then I'll try it from the outside.  THANK YOU FOR SAVING THIS HARDWARE.

**Comment by socelillyhammer** [comment]
@lepmail is your Verizon Fios Box the Arris IPC4100 with this remote? And if so has your remote coding worked with the bluetooth?
I cannot seem to change the channels with the remotes provided and was wondering if anyone can help here.

![rc4100_v2](https://user-images.githubusercontent.com/119097385/204646747-071b0318-633f-42c0-a78b-084d05648b77.png)

**Reply by monisallam** [reply]
+1 ^

**Reply by lepmail** [reply]
I'm afraid not @socelillyhammer. My box is the Cisco CHS435, but I believe this "FIOS" remote will work with your box too.  Unfortunately, I don't have any way to test that.  There should be some remote codes listed in your log when you run the slingbox_server.  Do you see them?

**Reply by socelillyhammer** [reply]
Yup I do see those codes.  But they dont work.

**Reply by mdehrlich** [reply]
If the STB expects bluetooth, it is never going to work with slinger remote.  Did slingbox control your STB before the changeover?  Slingboxes only do ir control.

**Reply by socelillyhammer** [reply]
@mdehrlich yes SB did control the STB up until SB shut down their servers and I couldnt log in to the webplayer any more.
